File: sdk/javascript/littleapi.js
/* LittleAPI JavaScript SDK
 * Browser / Node ESM client for https://littleapi.online
 * Source distribution. No npm publication is implied by this file.
 */

const DEFAULT_BASE_URL = "https://littleapi.online/api/v1";

function queryString(params = {}) {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params || {})) {
    if (value === undefined || value === null || value === "") continue;
    if (Array.isArray(value)) value.forEach(item => search.append(key, String(item)));
    else if (typeof value === "boolean") search.set(key, value ? "true" : "false");
    else search.set(key, String(value));
  }
  const text = search.toString();
  return text ? `?${text}` : "";
}

export class LittleAPIError extends Error {
  constructor(message, { status = 0, code = "request_failed", details = null, response = null } = {}) {
    super(message);
    this.name = "LittleAPIError";
    this.status = status;
    this.code = code;
    this.details = details;
    this.response = response;
  }
}

export class LittleAPI {
  constructor({ apiId, apiKey = "", baseUrl = DEFAULT_BASE_URL, fetchImpl = globalThis.fetch } = {}) {
    if (!apiId) throw new TypeError("apiId is required");
    if (typeof fetchImpl !== "function") throw new TypeError("A fetch implementation is required");
    this.apiId = String(apiId);
    this.apiKey = String(apiKey || "");
    this.baseUrl = String(baseUrl || DEFAULT_BASE_URL).replace(/\/$/, "");
    this.fetch = fetchImpl;
  }

  url(path = "", params = {}) {
    const suffix = String(path || "").replace(/^\/+/, "");
    return `${this.baseUrl}/${encodeURIComponent(this.apiId)}${suffix ? `/${suffix}` : ""}${queryString(params)}`;
  }

  async request(path = "", { method = "GET", params = {}, body, headers = {}, responseType = "json" } = {}) {
    const requestHeaders = new Headers(headers);
    if (this.apiKey && !requestHeaders.has("X-API-Key")) requestHeaders.set("X-API-Key", this.apiKey);
    let payload = body;
    if (body !== undefined && body !== null && !(body instanceof FormData) && !(body instanceof Blob) && !(body instanceof ArrayBuffer)) {
      if (!requestHeaders.has("Content-Type")) requestHeaders.set("Content-Type", "application/json");
      payload = JSON.stringify(body);
    }
    const response = await this.fetch(this.url(path, params), { method, headers: requestHeaders, body: payload });
    if (!response.ok) {
      let data = null;
      try { data = await response.clone().json(); } catch (_) { try { data = await response.clone().text(); } catch (_) {} }
      throw new LittleAPIError(data?.message || data?.error || `LittleAPI returned HTTP ${response.status}`, {
        status: response.status,
        code: data?.error || "request_failed",
        details: data?.details ?? null,
        response,
      });
    }
    if (responseType === "response") return response;
    if (responseType === "text") return response.text();
    if (responseType === "blob") return response.blob();
    if (responseType === "arrayBuffer") return response.arrayBuffer();
    if (response.status === 204) return null;
    return response.json();
  }

  // ----- Google Sheets: reads -----
  list(options = {}) { return this.request("", { params: options }); }
  search(filters = {}, options = {}) { return this.request("search", { params: { ...options, ...filters } }); }
  searchOr(filters = {}, options = {}) { return this.request("search_or", { params: { ...options, ...filters } }); }
  columns({ sheet } = {}) { return this.request("keys", { params: { sheet } }); }
  count({ sheet } = {}) { return this.request("count", { params: { sheet } }); }
  cells(coordinates, { sheet } = {}) {
    const value = Array.isArray(coordinates) ? coordinates.join(",") : String(coordinates);
    return this.request(`cells/${encodeURIComponent(value)}`, { params: { sheet } });
  }
  stats(column = "", { sheet } = {}) { return this.request("stats", { params: { sheet, column } }); }
  metadata() { return this.request("metadata"); }
  openapi() { return this.request("openapi.json"); }
  usage() { return this.request("usage"); }

  async query(query, options = {}) {
    const { sheet, headers = 1, columns = "names", raw = false, method = "auto" } = options;
    const usePost = method === "POST" || (method === "auto" && String(query).length > 1200);
    if (usePost) return this.request("query", { method: "POST", body: { query, sheet, headers, columns, raw } });
    return this.request("query", { params: { q: query, sheet, headers, columns, raw } });
  }

  // ----- Google Sheets: writes -----
  insert(rows, { sheet } = {}) { return this.request("", { method: "POST", params: { sheet }, body: rows }); }
  update(where, data, { sheet, casesensitive = false, all = false } = {}) {
    return this.request("", { method: "PATCH", params: { sheet }, body: { ...(all ? { all: true } : { where }), data, casesensitive } });
  }
  updateRange(range, values, { sheet } = {}) { return this.request("", { method: "PATCH", params: { sheet }, body: { range, values } }); }
  deleteRows(where, { sheet, casesensitive = false, all = false } = {}) {
    return this.request("", { method: "DELETE", params: { sheet }, body: { ...(all ? { all: true } : { where }), casesensitive } });
  }
  clear(range, { sheet } = {}) { return this.request("clear", { method: "POST", body: { sheet, range } }); }
  format(range, format, { sheet, gridRange } = {}) {
    return this.request("format", { method: "PATCH", body: { sheet, ...(gridRange ? { gridRange } : { range }), ...format } });
  }
  batch(requests) { return this.request("batch", { method: "POST", body: { requests } }); }

  // ----- Google Sheets: tabs -----
  sheets() { return this.request("sheets"); }
  createSheet(name) { return this.request("sheets", { method: "POST", body: { name } }); }
  renameSheet(name, newName) { return this.request(`sheets/${encodeURIComponent(name)}`, { method: "PATCH", body: { name: newName } }); }
  deleteSheet(name) { return this.request(`sheets/${encodeURIComponent(name)}`, { method: "DELETE" }); }
  copySheet(sheet, destinationSpreadsheetId) {
    return this.request("sheets/copy", { method: "POST", body: { sheet, destination_spreadsheet_id: destinationSpreadsheetId } });
  }

  // ----- Exports -----
  exportJson(options = {}) { return this.request("export.json", { params: options }); }
  exportCsv({ sheet } = {}) { return this.request("export.csv", { params: { sheet }, responseType: "text" }); }
  exportXlsx() { return this.request("export.xlsx", { responseType: "arrayBuffer" }); }

  // ----- Google Drive -----
  driveFile(fileId = "") { return this.request(fileId ? encodeURIComponent(fileId) : ""); }
  driveChildren(folderId, options = {}) { return this.request(`children/${encodeURIComponent(folderId)}`, { params: options }); }
  driveSearch({ name, parentId, pageSize, pageToken, orderBy } = {}) {
    return this.request("search", { params: { name, parent_id: parentId, page_size: pageSize, page_token: pageToken, order_by: orderBy } });
  }
  driveQuota() { return this.request("quota"); }
  driveAbout() { return this.request("about"); }
  driveCreate({ name, mimeType = "application/octet-stream", parentId = "", parents, contentBase64 } = {}) {
    const path = parentId ? encodeURIComponent(parentId) : "";
    return this.request(path, { method: "POST", body: { name, mimeType, parents, ...(contentBase64 ? { content_base64: contentBase64 } : {}) } });
  }
  driveRename(fileId, name) { return this.request(encodeURIComponent(fileId), { method: "PATCH", body: { name } }); }
  driveDelete(fileId) { return this.request(encodeURIComponent(fileId), { method: "DELETE" }); }
  driveDownload(fileId) { return this.request(`download/${encodeURIComponent(fileId)}`, { responseType: "arrayBuffer" }); }
  driveExport(fileId, mimeType) { return this.request(`export/${encodeURIComponent(fileId)}`, { params: { mime_type: mimeType }, responseType: "arrayBuffer" }); }
}

export default LittleAPI;

File: sdk/python/littleapi.py
"""LittleAPI Python SDK.

Standard-library client for https://littleapi.online.
This source is not yet published to PyPI.
"""

from __future__ import annotations

import json
from typing import Any, Dict, Iterable, Mapping, Optional
from urllib.parse import urlencode, quote
from urllib.request import Request, urlopen
from urllib.error import HTTPError

DEFAULT_BASE_URL = "https://littleapi.online/api/v1"


class LittleAPIError(RuntimeError):
    def __init__(self, message: str, *, status: int = 0, code: str = "request_failed", details: Any = None):
        super().__init__(message)
        self.status = status
        self.code = code
        self.details = details


def _params(values: Optional[Mapping[str, Any]]) -> str:
    pairs = []
    for key, value in (values or {}).items():
        if value is None or value == "":
            continue
        if isinstance(value, (list, tuple)):
            pairs.extend((key, str(item)) for item in value)
        elif isinstance(value, bool):
            pairs.append((key, "true" if value else "false"))
        else:
            pairs.append((key, str(value)))
    return urlencode(pairs, doseq=True)


class LittleAPI:
    def __init__(self, api_id: str, api_key: str = "", base_url: str = DEFAULT_BASE_URL, timeout: int = 30):
        if not api_id:
            raise ValueError("api_id is required")
        self.api_id = str(api_id)
        self.api_key = str(api_key or "")
        self.base_url = str(base_url or DEFAULT_BASE_URL).rstrip("/")
        self.timeout = timeout

    def url(self, path: str = "", params: Optional[Mapping[str, Any]] = None) -> str:
        suffix = str(path or "").lstrip("/")
        url = f"{self.base_url}/{quote(self.api_id, safe='')}"
        if suffix:
            url += "/" + suffix
        query = _params(params)
        return url + (("?" + query) if query else "")

    def request(self, path: str = "", *, method: str = "GET", params: Optional[Mapping[str, Any]] = None,
                body: Any = None, headers: Optional[Mapping[str, str]] = None, response_type: str = "json") -> Any:
        request_headers: Dict[str, str] = {
            "User-Agent": "LittleAPI-Python/1.0",
            "Accept": "application/json, text/plain;q=0.9, */*;q=0.8",
            **dict(headers or {}),
        }
        if self.api_key and "X-API-Key" not in request_headers:
            request_headers["X-API-Key"] = self.api_key
        payload = None
        if body is not None:
            payload = json.dumps(body, ensure_ascii=False).encode("utf-8")
            request_headers.setdefault("Content-Type", "application/json")
        req = Request(self.url(path, params), data=payload, headers=request_headers, method=method)
        try:
            with urlopen(req, timeout=self.timeout) as response:
                raw = response.read()
                if response_type == "bytes":
                    return raw
                text = raw.decode("utf-8", errors="replace")
                if response_type == "text":
                    return text
                return json.loads(text) if text else None
        except HTTPError as error:
            raw = error.read()
            text = raw.decode("utf-8", errors="replace")
            try:
                data = json.loads(text)
            except Exception:
                data = {}
            raise LittleAPIError(
                data.get("message") or data.get("error") or f"LittleAPI returned HTTP {error.code}",
                status=error.code,
                code=data.get("error", "request_failed"),
                details=data.get("details"),
            ) from error

    # Google Sheets reads
    def list(self, **options: Any) -> Any:
        return self.request(params=options)

    def search(self, filters: Optional[Mapping[str, Any]] = None, **options: Any) -> Any:
        return self.request("search", params={**options, **dict(filters or {})})

    def search_or(self, filters: Optional[Mapping[str, Any]] = None, **options: Any) -> Any:
        return self.request("search_or", params={**options, **dict(filters or {})})

    def columns(self, sheet: Optional[str] = None) -> Any:
        return self.request("keys", params={"sheet": sheet})

    def count(self, sheet: Optional[str] = None) -> Any:
        return self.request("count", params={"sheet": sheet})

    def cells(self, coordinates: Iterable[str] | str, sheet: Optional[str] = None) -> Any:
        value = ",".join(coordinates) if not isinstance(coordinates, str) else coordinates
        return self.request("cells/" + quote(value, safe=","), params={"sheet": sheet})

    def stats(self, column: str = "", sheet: Optional[str] = None) -> Any:
        return self.request("stats", params={"sheet": sheet, "column": column})

    def metadata(self) -> Any:
        return self.request("metadata")

    def openapi(self) -> Any:
        return self.request("openapi.json")

    def usage(self) -> Any:
        return self.request("usage")

    def query(self, query: str, *, sheet: Optional[str] = None, headers: int = 1,
              columns: str = "names", raw: bool = False, method: str = "auto") -> Any:
        use_post = method.upper() == "POST" or (method.lower() == "auto" and len(str(query)) > 1200)
        if use_post:
            return self.request("query", method="POST", body={"query": query, "sheet": sheet, "headers": headers, "columns": columns, "raw": raw})
        return self.request("query", params={"q": query, "sheet": sheet, "headers": headers, "columns": columns, "raw": raw})

    # Google Sheets writes
    def insert(self, rows: Any, sheet: Optional[str] = None) -> Any:
        return self.request(method="POST", params={"sheet": sheet}, body=rows)

    def update(self, where: Optional[Mapping[str, Any]], data: Mapping[str, Any], *, sheet: Optional[str] = None,
               casesensitive: bool = False, all_rows: bool = False) -> Any:
        body: Dict[str, Any] = {"data": dict(data), "casesensitive": casesensitive}
        body.update({"all": True} if all_rows else {"where": dict(where or {})})
        return self.request(method="PATCH", params={"sheet": sheet}, body=body)

    def update_range(self, range_a1: str, values: list, sheet: Optional[str] = None) -> Any:
        return self.request(method="PATCH", params={"sheet": sheet}, body={"range": range_a1, "values": values})

    def delete_rows(self, where: Optional[Mapping[str, Any]] = None, *, sheet: Optional[str] = None,
                    casesensitive: bool = False, all_rows: bool = False) -> Any:
        body: Dict[str, Any] = {"casesensitive": casesensitive}
        body.update({"all": True} if all_rows else {"where": dict(where or {})})
        return self.request(method="DELETE", params={"sheet": sheet}, body=body)

    def clear(self, range_a1: str, sheet: Optional[str] = None) -> Any:
        return self.request("clear", method="POST", body={"sheet": sheet, "range": range_a1})

    def format(self, range_a1: str, format_options: Mapping[str, Any], *, sheet: Optional[str] = None,
               grid_range: Optional[Mapping[str, Any]] = None) -> Any:
        body = {"sheet": sheet, **dict(format_options)}
        if grid_range is not None:
            body["gridRange"] = dict(grid_range)
        else:
            body["range"] = range_a1
        return self.request("format", method="PATCH", body=body)

    def batch(self, requests: list) -> Any:
        return self.request("batch", method="POST", body={"requests": requests})

    # Sheet tabs
    def sheets(self) -> Any:
        return self.request("sheets")

    def create_sheet(self, name: str) -> Any:
        return self.request("sheets", method="POST", body={"name": name})

    def rename_sheet(self, name: str, new_name: str) -> Any:
        return self.request("sheets/" + quote(name, safe=""), method="PATCH", body={"name": new_name})

    def delete_sheet(self, name: str) -> Any:
        return self.request("sheets/" + quote(name, safe=""), method="DELETE")

    def copy_sheet(self, sheet: str, destination_spreadsheet_id: str) -> Any:
        return self.request("sheets/copy", method="POST", body={"sheet": sheet, "destination_spreadsheet_id": destination_spreadsheet_id})

    # Exports
    def export_json(self, **options: Any) -> Any:
        return self.request("export.json", params=options)

    def export_csv(self, sheet: Optional[str] = None) -> str:
        return self.request("export.csv", params={"sheet": sheet}, response_type="text")

    def export_xlsx(self) -> bytes:
        return self.request("export.xlsx", response_type="bytes")

    # Google Drive
    def drive_file(self, file_id: str = "") -> Any:
        return self.request(quote(file_id, safe="") if file_id else "")

    def drive_children(self, folder_id: str, **options: Any) -> Any:
        return self.request("children/" + quote(folder_id, safe=""), params=options)

    def drive_search(self, *, name: Optional[str] = None, parent_id: Optional[str] = None,
                     page_size: Optional[int] = None, page_token: Optional[str] = None,
                     order_by: Optional[str] = None) -> Any:
        return self.request("search", params={"name": name, "parent_id": parent_id, "page_size": page_size, "page_token": page_token, "order_by": order_by})

    def drive_quota(self) -> Any:
        return self.request("quota")

    def drive_about(self) -> Any:
        return self.request("about")

    def drive_create(self, name: str, *, mime_type: str = "application/octet-stream", parent_id: str = "",
                     parents: Optional[list] = None, content_base64: Optional[str] = None) -> Any:
        path = quote(parent_id, safe="") if parent_id else ""
        body: Dict[str, Any] = {"name": name, "mimeType": mime_type}
        if parents is not None:
            body["parents"] = parents
        if content_base64:
            body["content_base64"] = content_base64
        return self.request(path, method="POST", body=body)

    def drive_rename(self, file_id: str, name: str) -> Any:
        return self.request(quote(file_id, safe=""), method="PATCH", body={"name": name})

    def drive_delete(self, file_id: str) -> Any:
        return self.request(quote(file_id, safe=""), method="DELETE")

    def drive_download(self, file_id: str) -> bytes:
        return self.request("download/" + quote(file_id, safe=""), response_type="bytes")

    def drive_export(self, file_id: str, mime_type: str) -> bytes:
        return self.request("export/" + quote(file_id, safe=""), params={"mime_type": mime_type}, response_type="bytes")
